src/Flysystem/FlysystemAssetStore.php

```diff
@@ -152,30 +179,177 @@ public function setProtectedFilesystem(Filesystem $filesystem)
+    }
+
     /**
      * Return the store that contains the given fileID
      *
      * @param string $fileID Internal file identifier
+     * @deprecated 1.4.0
      * @return Filesystem
      */
     protected function getFilesystemFor($fileID)
     {
-        if ($this->getPublicFilesystem()->has($fileID)) {
-            return $this->getPublicFilesystem();
+        return $this->applyToFileOnFilesystem(
+            function (ParsedFileID $parsedFileID, Filesystem $fs) {
+                return $fs;
+            },
+            $fileID
+        );
+    }
+
+    /**
+     * Generic method to apply an action to a file regardless of what FileSystem it's on. The action to perform should
+     * be provided as a closure expecting the following signature:
+     * ```
+     * function(ParsedFileID $parsedFileID, FileSystem $fs, FileResolutionStrategy $strategy, $visibility)
+     * ```
+     *
+     * `applyToFileOnFilesystem` will try to following steps and call the closure if they are succesfull:
+     * 1. Look for the file on the public filesystem using the explicit fileID provided.
+     * 2. Look for the file on the protected filesystem using the explicit fileID provided.
+     * 3. Look for the file on the public filesystem using the public resolution strategy.
+     * 4. Look for the file on the protected filesystem using the protected resolution strategy.
+     *
+     * If the closure returns `false`, `applyToFileOnFilesystem` will carry on and try the follow up steps.
+     *
+     * Any other value the closure returns (including `null`) will be returned to the calling function.
+     *
+     * @param callable $closure Action to apply.
+     * @param string|array|ParsedFileID $fileID File identication. Can be a string, a file tuple or a ParsedFileID
+     * @param bool $strictHashCheck
+     * @return mixed
+     */
+    private function applyToFileOnFilesystem(callable $closure, $fileID, $strictHashCheck = true)
+    {
+        $publicSet = [
+            $this->getPublicFilesystem(),
+            $this->getPublicResolutionStrategy(),
+            self::VISIBILITY_PUBLIC
+        ];
+
+        $protectedSet = [
+            $this->getProtectedFilesystem(),
+            $this->getProtectedResolutionStrategy(),
+            self::VISIBILITY_PROTECTED
+        ];
+
+        if (is_array($fileID)) {
+            $fileID = new ParsedFileID($fileID);
         }
 
-        if ($this->getProtectedFilesystem()->has($fileID)) {
-            return $this->getProtectedFilesystem();
+        /** @var Filesystem $fs */
+        /** @var FileResolutionStrategy $strategy */
+        /** @var string $visibility */
+
+        // First we try to search for exact file id string match
+        foreach ([$publicSet, $protectedSet] as $set) {
+            list($fs, $strategy, $visibility) = $set;
+
+            // Get a FileID string based on the type of FileID
+            $fileIdStr = $fileID instanceof ParsedFileID ?
```

Comment:
Instead of all of these argument type checking, either do argument normalisation early in the method, or (more preferably, as I earlier suggested) limit the allowed arguments.

Reply:
Fix https://github.com/silverstripe/silverstripe-assets/pull/223#discussion_r276498373